### PR TITLE
[move-package] Move Package resolution and compilation #139

### LINKED ISSUE #9155: [Bug][move-model] crash in ty.rs
# 🐛 Bug

Crash inside move-model/ty.rs while attempting to run prover on some StarCoin code for benchmarking

## To reproduce

**Code snippet to reproduce**
See attached repro.
[repro.zip](https://github.com/diem/diem/files/7156308/repro.zip)

```rust
unzip repro.zip
mvp crash_repro/starswap_modules crash_repro/starcoin_modules
```

**Stack trace/error message**
```
thread 'main' panicked at 'index out of bounds: the len is 1 but the index is 1', language/move-model/src/ty.rs:270:21
stack backtrace:
   0: rust_begin_unwind
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/std/src/panicking.rs:515:5
   1: core::panicking::panic_fmt
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:92:14
   2: core::panicking::panic_bounds_check
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/panicking.rs:69:5
   3: move_model::ty::Type::replace
             at ./language/move-model/src/ty.rs:270:21
   4: move_model::ty::Type::replace::{{closure}}::{{closure}}
             at ./language/move-model/src/ty.rs:266:65
   5: core::iter::adapters::map::map_fold::{{closure}}
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/adapters/map.rs:82:28
   6: core::iter::traits::iterator::Iterator::fold
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/traits/iterator.rs:2112:21
   7: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/adapters/map.rs:122:9
   8: core::iter::traits::iterator::Iterator::for_each
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/traits/iterator.rs:736:9
   9: <alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/vec/spec_extend.rs:40:17
  10: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/vec/spec_from_iter_nested.rs:56:9
  11: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/vec/spec_from_iter.rs:33:9
  12: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/vec/mod.rs:2449:9
  13: core::iter::traits::iterator::Iterator::collect
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/traits/iterator.rs:1748:9
  14: move_model::ty::Type::replace::{{closure}}
             at ./language/move-model/src/ty.rs:266:44
  15: move_model::ty::Type::replace
             at ./language/move-model/src/ty.rs:293:70
  16: move_model::ty::Type::replace::{{closure}}::{{closure}}
             at ./language/move-model/src/ty.rs:266:65
  17: core::iter::adapters::map::map_fold::{{closure}}
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/adapters/map.rs:82:28
  18: core::iter::traits::iterator::Iterator::fold
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/traits/iterator.rs:2112:21
  19: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::fold
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/adapters/map.rs:122:9
  20: core::iter::traits::iterator::Iterator::for_each
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/traits/iterator.rs:736:9
  21: <alloc::vec::Vec<T,A> as alloc::vec::spec_extend::SpecExtend<T,I>>::spec_extend
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/vec/spec_extend.rs:40:17
  22: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/vec/spec_from_iter_nested.rs:56:9
  23: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/vec/spec_from_iter.rs:33:9
  24: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/alloc/src/vec/mod.rs:2449:9
  25: core::iter::traits::iterator::Iterator::collect
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/iter/traits/iterator.rs:1748:9
  26: move_model::ty::Type::replace::{{closure}}
             at ./language/move-model/src/ty.rs:266:44
  27: move_model::ty::Type::replace
             at ./language/move-model/src/ty.rs:293:70
  28: move_model::ty::Type::instantiate
             at ./language/move-model/src/ty.rs:233:13
  29: <move_model::spec_translator::SpecTranslator<T> as move_model::exp_rewriter::ExpRewriterFunctions>::rewrite_temporary
             at ./language/move-model/src/spec_translator.rs:628:34
  30: move_model::exp_rewriter::ExpRewriterFunctions::rewrite_exp_descent
             at ./language/move-model/src/exp_rewriter.rs:185:40
  31: <move_model::spec_translator::SpecTranslator<T> as move_model::exp_rewriter::ExpRewriterFunctions>::rewrite_exp
             at ./language/move-model/src/spec_translator.rs:582:19
  32: move_model::exp_rewriter::ExpRewriterFunctions::internal_rewrite_exp
             at ./language/move-model/src/exp_rewriter.rs:342:23
  33: move_model::exp_rewriter::ExpRewriterFunctions::internal_rewrite_vec
             at ./language/move-model/src/exp_rewriter.rs:352:30
  34: move_model::exp_rewriter::ExpRewriterFunctions::rewrite_exp_descent
             at ./language/move-model/src/exp_rewriter.rs:205:36
  35: <move_model::spec_translator::SpecTranslator<T> as move_model::exp_rewriter::ExpRewriterFunctions>::rewrite_exp
             at ./language/move-model/src/spec_translator.rs:582:19
  36: move_model::exp_rewriter::ExpRewriterFunctions::internal_rewrite_exp
             at ./language/move-model/src/exp_rewriter.rs:342:23
  37: move_model::exp_rewriter::ExpRewriterFunctions::internal_rewrite_vec
             at ./language/move-model/src/exp_rewriter.rs:361:32
  38: move_model::exp_rewriter::ExpRewriterFunctions::rewrite_exp_descent
             at ./language/move-model/src/exp_rewriter.rs:205:36
  39: <move_model::spec_translator::SpecTranslator<T> as move_model::exp_rewriter::ExpRewriterFunctions>::rewrite_exp
             at ./language/move-model/src/spec_translator.rs:582:19
  40: move_model::exp_rewriter::ExpRewriterFunctions::internal_rewrite_exp
             at ./language/move-model/src/exp_rewriter.rs:342:23
  41: move_model::exp_rewriter::ExpRewriterFunctions::internal_rewrite_vec
             at ./language/move-model/src/exp_rewriter.rs:360:32
  42: move_model::exp_rewriter::ExpRewriterFunctions::rewrite_exp_descent
             at ./language/move-model/src/exp_rewriter.rs:205:36
  43: <move_model::spec_translator::SpecTranslator<T> as move_model::exp_rewriter::ExpRewriterFunctions>::rewrite_exp
             at ./language/move-model/src/spec_translator.rs:582:19
  44: move_model::spec_translator::SpecTranslator<T>::translate_exp
             at ./language/move-model/src/spec_translator.rs:473:9
  45: move_model::spec_translator::SpecTranslator<T>::translate_spec
             at ./language/move-model/src/spec_translator.rs:355:17
  46: move_model::spec_translator::SpecTranslator<T>::translate_fun_spec
             at ./language/move-model/src/spec_translator.rs:225:9
  47: bytecode::spec_instrumentation::Instrumenter::instrument_call
             at ./language/move-prover/bytecode/src/spec_instrumentation.rs:496:31
  48: bytecode::spec_instrumentation::Instrumenter::instrument_bytecode
             at ./language/move-prover/bytecode/src/spec_instrumentation.rs:455:17
  49: bytecode::spec_instrumentation::Instrumenter::instrument
             at ./language/move-prover/bytecode/src/spec_instrumentation.rs:380:13
  50: bytecode::spec_instrumentation::Instrumenter::run
             at ./language/move-prover/bytecode/src/spec_instrumentation.rs:265:9
  51: <bytecode::spec_instrumentation::SpecInstrumentationProcessor as bytecode::function_target_pipeline::FunctionTargetProcessor>::process
             at ./language/move-prover/bytecode/src/spec_instrumentation.rs:120:33
  52: bytecode::function_target_pipeline::FunctionTargetProcessor::process_and_maybe_remove
             at ./language/move-prover/bytecode/src/function_target_pipeline.rs:96:14
  53: bytecode::function_target_pipeline::FunctionTargetsHolder::process
             at ./language/move-prover/bytecode/src/function_target_pipeline.rs:263:43
  54: bytecode::function_target_pipeline::FunctionTargetPipeline::run_with_hook
             at ./language/move-prover/bytecode/src/function_target_pipeline.rs:370:21
  55: bytecode::function_target_pipeline::FunctionTargetPipeline::run
             at ./language/move-prover/bytecode/src/function_target_pipeline.rs:380:9
  56: move_prover::create_and_process_bytecode
             at ./language/move-prover/src/lib.rs:239:9
  57: move_prover::run_move_prover_with_model
             at ./language/move-prover/src/lib.rs:114:19
  58: move_prover::run_move_prover
             at ./language/move-prover/src/lib.rs:58:5
  59: move_prover::run
             at ./language/move-prover/src/main.rs:33:5
  60: move_prover::main
             at ./language/move-prover/src/main.rs:12:21
  61: core::ops::function::FnOnce::call_once
             at /rustc/a178d0322ce20e33eac124758e837cbd80a6f633/library/core/src/ops/function.rs:227:5
```

## System information

Using latest Diem, version d2a6c5c0017cc5edcec82046b8a1baae41378393

### LINKED ISSUE #9167: [Bug][move-prover] unexpected result on global invariant
# 🐛 Bug

I would expect the prover to report an invariant violation on the following code, but it does not:

```
module 0x1::StructEq {
    struct S { f: u64 }

    invariant forall s: S: s == S { f: 10 };

    public fun new(): S {
        S { f: 11 }
    }
}
```

It's possible that the spec doesn't mean what I think it does. I want to say "all struct `S` values have `f` fields equal to 10. The more idiomatic way to write this is is `spec S { invariant f == 10 }`--the prover does correctly flag this as violated for the above code). I wrote the invariant in this funky way as a stress-test for the escape analysis, which tries to determine if leaked mutable references to fields are relevant to the module's specs.

### PR DESCRIPTION
**Pull request by satyamacn**
### Motivation

The bottom commit is resolution of packages, and the top commit adds compilation of packages.

This PR adds package resolution and compilation of packages for the Move Package system. This:

Parses and verifies that all dependencies exist (and their locations)
Instantiates all named addresses, and makes sure that their values agree across the dependency graph (more on this at the bottom).
Builds the dependency graph of packages, and annotates each package with the named addresses that are in scope in that package, their values, and the renamings that have been performed in that package. It also checks for cyclic dependencies and returns the offending cycle if one is found.
Naming
A named address can be assigned a value a couple different ways:

It can be assigned in the package that declares it. In this case that named address will always be that value.
It can be unassigned in the package that declares it. In this case another package that depends on it (transitively) can assign it a value. Multiple packages can assign a value to the same named address as long as those values are the same. The named address's value must be the same across the entire dependency graph.
Renaming
Addresses can be renamed from a package SomeDep when it is used as a dependency in a package P:

SomeDep = { local = "../path", addr_subst = { "SomeDepA" = "A" } } # Rename "A" from SomeDep to SomeDepA
After this, all packages that depend on P will not see the named address A in the names that are available to it (unless P redeclares A) and SomeDepA will be in the set of named addresses in scope.

### Test Plan
- CI/CD Tests are covered

------------------------
### Motivation
This PR adds a folder that contains what you need to start a Diem node locally with a custom genesis module set, along with
some sample Move code and a sample application sending transactions to the local validator node you just started.

The README contains more detailed information on what the code does and how to use it.

### Test Plan 
- CI/CD Tests are covered

-----------------------

### Motivation 

Builds on #8775 to add documentation and ABI generation to the package system. Additionally, in preparation for hooking up the Move Prover with packages, model building for a whole package is added. Only look at the top commit.

It is important to note that because the Move Model requires source files, at this time we cannot support renaming of named addresses in packages where you need to build a Move Model for a package and all of its dependencies. In the future we will want to support this, however this will most likely require support from the source language or compiler (to either support renaming at expansion, or in generation of interface files).

### Test Plan 
- CI/CD Tests are covered

-----------------------
### Motivation
This PR builds on PR #9030 and adds CLI support for the package system. It adds support for three basic Move Package commands to the CLI. Currently, these are:

move package build: build the package
move package errmap: build an errmap for the package (and its dependencies)
move package prove: run the Move Prover on the package. Note that the package is currently subject to restrictions in [move-package] Add support for documentation and ABI generation and building of Move Model.  #9030. Prover flags can be passed through this command as well.
This PR does not add unit testing support, that will be a separate PR.

In the process of adding support the Move CLI lib.rs was getting a bit hard to follow, so I refactored the CLI structure so that each subcommand module holds its CLI args and argument handler. This PR does not convert the CLI over to using the package system, nor does it convert over the move-stdlib and diem-framework. These will be done in upcoming PRs.

### Test Plan 
- CI/CD Tests are covered

---------------------------------
### Motivation
Simplify the Move IR language and compiler by eliminating the need to keep track of the types of expressions.

For the most part, this requires no change to the language, except for one aspect: when struct fields are borrowed, the user must now explicitly specify which struct type is being accessed. This is because, for example, a simple field identifier f may refer to different fields in any number of different structs S1, S2, etc.

What's in this pull request?
The file count is large because this includes test changes -- the *.mvir files. You can look at them to see what the new field access syntax in Move IR looks like: s.S<T>::f, where s is a struct instance, S<T> is its type, and f is the name of the field being accessed.
The rest of the changes, in the *.rs files are almost exclusively just deleting the InferredType enum.

### Test Plan 
- CI/CD Tests are covered
--------------------

### Motivation

This re-activates and refines the specification variable implementation of the Move prover.

Spec variables are declared as previously, but with the option to add an initializer:

spec module {
   global spec_var: num = 1; // optional initializer
}
An update of a spec var can appear in a function spec block:

fun f() ...
spec f {
  ...
  update spec_var = spec_var + 1; // assignment; no old
}

In contrast to the older defunct implementation, this one maps spec vars to regular global "ghost" memory. That is for each global x: T, there is a generated struct Ghost$x has key { v: T }. Access to spec vars is right-away mapped to accessing this memory (at address @0) in move-model. This way, memory usage and global invariant analysis can be applied to spec vars uniformly.

For a number of examples, see the test global_vars.move.

This implementation isn't complete yet; there is some more work to do with opaque functions.

### Test Plan
- CI/CD Tests were performed. 

---------------------

### Motivation

[diem-framework] Port ValidatorConfig and ValidatorOperatorConfig to unit tests

Ported testing for the ValidatorConfig and ValidatorOperatorConfig tests to be unit tests.

### Test Plan

- CI/CD Tests are covered

### Related PR
#9131

------------------------------

### Motivation

[language][diem-framework] migrate registered currency tests 

This migrates tests in move-lang/functional-tests/diem/registered_currencies into diem-framework and rewrite them as unit tests.

### Test Plan 
- CI/CD tests are covered

#Related PR
#9115

-----------------------------

### Motivation

[language][diem-framework] migrate authenticator tests
This PR migrates move-lang/functional-tests/diem/authenticator tests into diem-framework and rewrite them as unit tests.

### Test Plan
- CI/CD Tests are covered

### Related PR
#9114

--------------------------------

### Motivation

If the prover finds that a global invariant is not checked
in any function, throw a warning for that invariant.
The the new unused_global_invariant.move test case for the effect of
this checking.
This fixes #9167.
This warning is good to have and helps filter out unused specs.

### Test Plan
- In CI/CD, new test cases were added

--------------------------------------------------
### Motivation

Simplified the logic there a bit.

This PR fixes #9155, now the starswap_modules and starcoin_modules verify without a problem.

A minimal repro is produced under regression/temporary_rewrite.move.

### Test Plan
- CI/CD tests were performed

-----------------------------------------------------

### CODE REVIEW

**Review by dhaneshacn — APPROVED** [review]

**Comment by satyamacn** [comment]
Closing this PR due to technical issue. Raised new PR- https://github.com/diem/diem/pull/10698
